src/librustc_builtin_macros/asm.rs

```diff
@@ -0,0 +1,527 @@
+            if s == "asm!" {
+                err.span_suggestion(
+                    asm_span,
+                    "replace with",
+                    "llvm_asm!".into(),
+                    Applicability::MachineApplicable,
+                );
+            }
+        }
+        return Err(err);
+    }
+
+    let template = p.parse_expr()?;
+    let mut args = AsmArgs {
+        template,
+        operands: vec![],
+        named_args: FxHashMap::default(),
+        reg_args: FxHashSet::default(),
+        options: InlineAsmOptions::empty(),
+        options_span: None,
+    };
+
+    let mut first = true;
+    while p.token != token::Eof {
+        if !p.eat(&token::Comma) {
+            if first {
+                // After `asm!(""` we always expect *only* a comma...
+                let mut err = ecx.struct_span_err(p.token.span, "expected token: `,`");
+                err.span_label(p.token.span, "expected `,`");
+                p.maybe_annotate_with_ascription(&mut err, false);
+                return Err(err);
+            } else {
+                // ...after that delegate to `expect` to also include the other expected tokens.
+                return Err(p.expect(&token::Comma).err().unwrap());
+            }
+        }
+        first = false;
+        if p.token == token::Eof {
+            break;
+        } // accept trailing commas
+
+        let span_start = p.token.span;
+
+        // Parse options
+        if p.eat(&token::Ident(sym::options, false)) {
+            p.expect(&token::OpenDelim(token::DelimToken::Paren))?;
+
+            while !p.eat(&token::CloseDelim(token::DelimToken::Paren)) {
+                if p.eat(&token::Ident(sym::pure, false)) {
+                    args.options |= InlineAsmOptions::PURE;
+                } else if p.eat(&token::Ident(sym::nomem, false)) {
+                    args.options |= InlineAsmOptions::NOMEM;
+                } else if p.eat(&token::Ident(sym::readonly, false)) {
+                    args.options |= InlineAsmOptions::READONLY;
+                } else if p.eat(&token::Ident(sym::preserves_flags, false)) {
+                    args.options |= InlineAsmOptions::PRESERVES_FLAGS;
+                } else if p.eat(&token::Ident(sym::noreturn, false)) {
+                    args.options |= InlineAsmOptions::NORETURN;
+                } else {
+                    p.expect(&token::Ident(sym::nostack, false))?;
+                    args.options |= InlineAsmOptions::NOSTACK;
+                }
+
+                // Allow trailing commas
+                if p.eat(&token::CloseDelim(token::DelimToken::Paren)) {
+                    break;
+                }
+                p.expect(&token::Comma)?;
+            }
+
+            let new_span = span_start.to(p.prev_token.span);
+            if let Some(options_span) = args.options_span {
+                ecx.struct_span_err(new_span, "asm options cannot be specified twice")
```

Comment:
`s/twice/multiple times/` as user could specify options three or four or five… times as well. The error message should make sense in all of those instances.